core/matrix/permutation.cpp

```diff
@@ -33,8 +33,259 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+                                    const dim<2>& size)
+    : Permutation{exec, size[0]}
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(size);
+}
+
+
+template <typename IndexType>
+Permutation<IndexType>::Permutation(std::shared_ptr<const Executor> exec,
+                                    const dim<2>& size,
+                                    const mask_type& enabled_permute)
+    : Permutation{exec, size[0]}
+{
+    GKO_ASSERT_EQ(enabled_permute, row_permute);
+    GKO_ASSERT_IS_SQUARE_MATRIX(size);
+}
+
+
+template <typename IndexType>
+size_type Permutation<IndexType>::get_permutation_size() const noexcept
+{
+    return this->get_size()[0];
+}
+
+
+template <typename IndexType>
+mask_type Permutation<IndexType>::get_permute_mask() const
+{
+    return row_permute;
+}
+
+
+template <typename IndexType>
+void Permutation<IndexType>::set_permute_mask(mask_type permute_mask)
+{
+    GKO_ASSERT_EQ(permute_mask, row_permute);
+}
+
+
+template <typename IndexType>
+std::unique_ptr<Permutation<IndexType>> Permutation<IndexType>::invert() const
+{
+    const auto exec = this->get_executor();
+    const auto size = this->get_size()[0];
+    auto result = Permutation<IndexType>::create(exec, size);
+    exec->run(permutation::make_invert(this->get_const_permutation(), size,
+                                       result->get_permutation()));
+    return result;
+}
+
+
+template <typename IndexType>
+std::unique_ptr<Permutation<IndexType>> Permutation<IndexType>::combine(
+    ptr_param<const Permutation<IndexType>> other) const
+{
+    GKO_ASSERT_EQUAL_DIMENSIONS(this, other);
+    const auto exec = this->get_executor();
+    const auto size = this->get_size()[0];
+    const auto local_other = make_temporary_clone(exec, other);
+    auto result = Permutation<IndexType>::create(exec, size);
+    exec->run(permutation::make_combine(this->get_const_permutation(),
+                                        local_other->get_const_permutation(),
+                                        size, result->get_permutation()));
+    return result;
+}
+
+
+template <typename IndexType>
+void Permutation<IndexType>::write(
+    gko::matrix_data<value_type, index_type>& data) const
+{
+    const auto host_this =
+        make_temporary_clone(this->get_executor()->get_master(), this);
+    data.size = this->get_size();
+    data.nonzeros.clear();
+    data.nonzeros.reserve(data.size[0]);
+    for (IndexType row = 0; row < this->get_size()[0]; row++) {
+        data.nonzeros.emplace_back(row, host_this->get_const_permutation()[row],
+                                   one<value_type>());
+    }
+}
+
+
+template <typename Functor>
+void dispatch_dense(const LinOp* op, Functor fn)
+{
+    using matrix::Dense;
+    using std::complex;
+    run<const Dense<double>*, const Dense<float>*,
+        const Dense<complex<double>>*, const Dense<complex<float>>*>(op, fn);
+}
+
+
+template <typename IndexType>
+void Permutation<IndexType>::apply_impl(const LinOp* in, LinOp* out) const
+{
```

Comment:
Can we maybe have a check here that throws if the input LinOp is not Dense ? With maybe more information that the user should instead use the class provided `permute` function ? 

Because from a user point of view, applying a permutation to any LinOp makes sense.